apps/emqx/src/emqx_log_throttler.erl

```diff
@@ -0,0 +1,139 @@
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%--------------------------------------------------------------------
+
+-module(emqx_log_throttler).
+
+-behaviour(gen_server).
+
+-include("logger.hrl").
+-include("types.hrl").
+-include_lib("snabbkaffe/include/snabbkaffe.hrl").
+
+-export([start_link/0]).
+
+%% throttler API
+-export([allow/2]).
+
+%% gen_server callbacks
+-export([
+    init/1,
+    handle_call/3,
+    handle_cast/2,
+    handle_info/2,
+    terminate/2,
+    code_change/3
+]).
+
+-define(SEQ_ID(EventId), {?MODULE, EventId}).
+-define(NEW_SEQ, atomics:new(1, [{signed, false}])).
+-define(GET_SEQ(EventId), persistent_term:get(?SEQ_ID(EventId), undefined)).
+-define(RESET_SEQ(SeqRef), atomics:put(SeqRef, 1, 0)).
+-define(INC_SEQ(SeqRef), atomics:add(SeqRef, 1, 1)).
+-define(GET_DROPPED(SeqRef), atomics:get(SeqRef, 1) - 1).
+-define(IS_ALLOWED(SeqRef), atomics:add_get(SeqRef, 1, 1) =:= 1).
+
+-define(NEW_THROTTLE(EventId, SeqRef), persistent_term:put(?SEQ_ID(EventId), SeqRef)).
+
+-define(EVENT_IDS_LIST, emqx:get_config([log, throttling, event_ids], [])).
+-define(WINDOW_TIME_SEC, emqx:get_config([log, throttling, window_time], 60)).
+
+-spec allow(logger:level(), atom()) -> boolean().
+allow(debug, _EventId) ->
+    true;
+allow(_Level, EventId) when is_atom(EventId) ->
+    Seq = persistent_term:get(?SEQ_ID(EventId), undefined),
+    case Seq of
+        undefined ->
+            %% This is either a race condition (emqx_log_throttler is not started yet)
+            %% or a developer mistake (event id used in ?SLOG_THROTTLE/3,4 macro is
+            %% not added to the default value of `log.throttling.event_ids`.
+            ?SLOG(info, #{
+                msg => "missing_log_throttle_sequence",
+                throttle_event_id => EventId
+            }),
+            true;
+        SeqRef ->
+            ?IS_ALLOWED(SeqRef)
+    end.
+
+-spec start_link() -> startlink_ret().
+start_link() ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
+
+%%--------------------------------------------------------------------
+%% gen_server callbacks
+%%--------------------------------------------------------------------
+
+init([]) ->
+    ok = lists:foreach(fun(Id) -> ?NEW_THROTTLE(Id, ?NEW_SEQ) end, ?EVENT_IDS_LIST),
+    TimerRef = schedule_refresh(?WINDOW_TIME_SEC),
+    {ok, #{timer_ref => TimerRef}}.
+
+handle_call(Req, _From, State) ->
+    ?SLOG(error, #{msg => "unexpected_call", call => Req}),
+    {reply, ignored, State}.
+
+handle_cast(Msg, State) ->
+    ?SLOG(error, #{msg => "unexpected_cast", cast => Msg}),
+    {noreply, State}.
+
+handle_info(refresh, State) ->
+    PeriodSec = ?WINDOW_TIME_SEC,
+    EventIds = ?EVENT_IDS_LIST,
+    ok = lists:foreach(
```

Comment:
maybe log a summary instead of one for each id?
e.g.
`dropped => #{id1 => 12, id2 => 34}`